domain/src/main/java/com/github/ianellis/shifts/domain/ShiftRepository.kt

```diff
@@ -1,67 +1,47 @@
 package com.github.ianellis.shifts.domain
 
 import com.github.ianellis.shifts.domain.events.Event
-import com.github.ianellis.shifts.domain.events.EventEmitter
 import com.github.ianellis.shifts.entities.EndShiftRequest
 import com.github.ianellis.shifts.entities.ISO8601
 import com.github.ianellis.shifts.entities.ShiftEntity
 import com.github.ianellis.shifts.entities.StartShiftRequest
-import kotlinx.coroutines.CoroutineDispatcher
-import kotlinx.coroutines.Deferred
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.GlobalScope
-import kotlinx.coroutines.async
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.ConflatedBroadcastChannel
 
+@ExperimentalCoroutinesApi
 class ShiftRepository(
     private val service: ShiftsService,
-    private val cache: ShiftsCache,
-    private val backgroundDispatcher: CoroutineDispatcher,
-    private val eventEmitter: EventEmitter<List<ShiftEntity>>
+    private val cache: ShiftsCache
 ) {
 
-    fun addListener(listener: (Event<List<ShiftEntity>>) -> Unit) {
```

Comment:
I am not sure the concept of channels applies here. Channels provide a way to transfer a stream of values. In this example we just observe changes, so in my opinion, LiveData fits here better.

**Streams vs Observers**
With the release of RxJava and influence of reactive programming, there is a popular opinion that _everything is a stream_. I don't think this is correct. People tend to use one tool to solve all the problem. It should be vice-versa, for every problem there is a tool.

Let's say we want to read content from the text file. This problem can be solved in different ways.

Example of solving the problem with a stream:
```kotlin
fun readLogFile(): Channel<String> {
val channel = Channel<String>()
File("log.txt").readLines { line:String ->
    channel.send(line) // send line one by one
}
return channel
```

Example of solving the problem with observer:
```kotlin
fun readLogFile(callback: (result: String) -> Unit) {
val builder = StringBuilder()
File("log.txt").readLines { line:String ->
    builder.add(line)
}
callback(builder.toString()) // send whole text
```

**Streams** must be used when we want to deliver the result by pieces and then manipulate it via different operations like map, filter, group, etc.

**Observers** must be used when we want to deliver complete result (can be periodic) and most of the time this observer is in idle state e.g. click listener, database table changes, etc.

Reply:
Interesting points, and probably a topic for discussion outside of Pr comment. However, Streams are simply an implementation of the Observer pattern that provide an API to transform the data, which is what I want. 

I specifically do not want to use LiveData as it is an Android specific platform integration. This tightly couples your architecture to the platform. 

Callbacks as we know do not scale well.

We need to find some API that models data that changes over time and allows us a way to transform and combine data sources. LiveData could possibly solve this but I would prefer a platform agnostic way to achieve this if possible